lnwallet/channel.go

```diff
@@ -7214,8 +7214,48 @@ func CreateCooperativeCloseTx(fundingTxIn wire.TxIn,
 	return closeTx
 }
 
-// CalcFee returns the commitment fee to use for the given
-// fee rate (fee-per-kw).
+// LocalBalanceDust returns true if when creating a co-op close transaction,
+// the balance of the local party will be dust after accounting for any anchor
+// outputs.
+func (lc *LightningChannel) LocalBalanceDust() bool {
+	lc.RLock()
+	defer lc.RUnlock()
+
+	chanState := lc.channelState
+	localBalance := chanState.LocalCommitment.LocalBalance.ToSatoshis()
+
+	// If this is an anchor channel, and we're the initiator, then we'll
+	// regain the stats allocated to the anchor outputs with the co-op
+	// close transaction.
+	if chanState.ChanType.HasAnchors() && chanState.IsInitiator {
+		localBalance += 2 * anchorSize
+	}
+
+	return localBalance <= chanState.LocalChanCfg.DustLimit
+}
+
+// RemoteBalanceDust returns true if when creating a co-op close transaction,
+// the balance of the remote party will be dust after accounting for any anchor
+// outputs.
+func (lc *LightningChannel) RemoteBalanceDust() bool {
+	lc.RLock()
+	defer lc.RUnlock()
+
+	chanState := lc.channelState
+	remoteBalance := chanState.RemoteCommitment.RemoteBalance.ToSatoshis()
+
+	// If this is an anchor channel, and they're the initiator, then we'll
+	// regain the stats allocated to the anchor outputs with the co-op
+	// close transaction.
+	if chanState.ChanType.HasAnchors() && !chanState.IsInitiator {
+		remoteBalance += 2 * anchorSize
+	}
+
+	return remoteBalance <= chanState.RemoteChanCfg.DustLimit
```

Comment:
If one side has dust, this can lead to coop close not working. One side should be able to broadcast in these scenarios, but the spec doesn't say what to do if you receive a `fee_satoshis` that makes the transaction non-standard. I only included the relevant scenarios, there are more that do not matter:

Scenario 1 :
| Commit | Our Output | Their Output |
| ------------- | ------------- | ------------- |
|  Local | dust  | not  |
|  Remote  | not  | not  |

Our output is dust on our commitment, so we will not include it as weight and will send a fee that may be too low to the counterparty. The counterparty does not see this output as dust. If the fee is low enough, it may be non-standard for the counterparty. 

Scenario 2:
| Commit | Our Output | Their Output |
| ------------- | ------------- | ------------- |
|  Local | not  | dust  |
|  Remote  | not  | not  |

Their output is dust on our commitment. We will still use the weight of the trimmed output and send them a fee that will be standard for their commitment transaction. This is ok.

Scenario 3: 
| Commit | Our Output | Their Output |
| ------------- | ------------- | ------------- |
|  Local | not  | not  |
|  Remote  | not  | dust  |

Their output is dust on their commitment, so we will not use their weight. We will propose a fee that is non-standard for us, but not for them. This isn't ok because if they accept it, we will not be able to broadcast the closing txn.

Scenario 4:
| Commit | Our Output | Their Output |
| ------------- | ------------- | ------------- |
|  Local | not  | not  |
|  Remote  | dust  | not  |

This is not dust on ours, but is on theirs. So we will include the weight when proposing the fee. This is fine.

It seems like the best way would be to not trim outputs if they are dust. Then you slightly overpay, but never underpay.

Reply:
> It seems like the best way would be to not trim outputs if they are dust. Then you slightly overpay, but never underpay.

You mean trim during fee estimation (here), or when we actually construct the co-op transaction? If we don't trim when making the co-op close transaction, and the other party does, then we'll get an invalid sig (diff sighash) and the fee divergence won't matter. 

I originally didn't do any trimming at all (for fee estimation) but then switched it in an attempt to "optimize" fees a bit more. However, just removing that logic does ensure we'll only ever slightly overpay and don't need to worry about the underway case (lack of mempool propagation).

Reply:
I meant not trimming for fee estimation, we'd have to trim the outputs according to spec to not get invalid sig